File: 0054.SpiralMatrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        ans = []
        while matrix:
            ans += matrix.pop(0)
            if matrix and matrix[0]:
                for row in matrix:
                    ans.append(row.pop())
            if matrix:
                ans += matrix.pop()[::-1]
            if matrix and matrix[0]:
                for row in matrix[::-1]:
                    ans.append(row.pop(0))
        return ans

        """
        if not matrix:
            return []
        ans = []
        t, b = 0, len(matrix)-1
        l, r = 0, len(matrix[0])-1
        while t < b and l < r:
            ans.extend([matrix[t][j] for j in xrange(l, r)])
            ans.extend([matrix[i][r] for i in xrange(t, b)])
            ans.extend([matrix[b][j] for j in xrange(r, l, -1)])
            ans.extend([matrix[i][l] for i in xrange(b, t, -1)])
            '''
            for j in xrange(l, r):
                ans.append(matrix[t][j])
            for i in xrange(t, b):
                ans.append(matrix[i][r])
            for j in xrange(r, l, -1):
                ans.append(matrix[b][j])
            for i in xrange(b, t, -1):
                ans.append(matrix[i][l])
            '''
            t += 1
            b -= 1
            l += 1
            r -= 1
        if t == b:
            ans.extend([matrix[t][j] for j in xrange(l, r+1)])
            '''
            for j in xrange(l, r+1):
                ans.append(matrix[t][j])
            '''
            #return ans
        elif l == r:
            ans.extend([matrix[i][l] for i in xrange(t, b+1)])
            '''
            for i in xrange(t, b+1):
                ans.append(matrix[i][l])
            '''
        return ans
        """

File: 0054.SpiralMatrix.go
func spiralOrder(matrix [][]int) []int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return []int{}
    }
    i1, i2 := 0, len(matrix)-1
    j1, j2 := 0, len(matrix[0])-1
    spiral := []int{}
    for i1 <= i2 && j1 <= j2 {
        for j := j1; j <= j2; j++ {
            spiral = append(spiral, matrix[i1][j])
        }
        for i := i1+1; i <= i2; i++ {
            spiral = append(spiral, matrix[i][j2])
        }
        if i1 < i2 && j1 < j2 {
            for j := j2-1; j >= j1; j-- {
                spiral = append(spiral, matrix[i2][j])
            }
            for i := i2-1; i > i1; i-- {
                spiral = append(spiral, matrix[i][j1])
            }
        }
        i1, i2 = i1+1, i2-1
        j1, j2 = j1+1, j2-1
    }
    return spiral
}